purchases/src/main/kotlin/com/revenuecat/purchases/PurchaseParams.kt

```
@@ -122,6 +128,64 @@ class PurchaseParams(val builder: Builder) {
             this.googleReplacementMode = googleReplacementMode
         }
 
+        /*
+         * The [Package]s to add on to the base package passed in via the [PurchaseParams.Builder]'s constructor.
+         * This will result in a multi-line purchase whose base product is the one passed in to the
+         * [PurchaseParams.Builder]'s constructor.
+         *
+         * The following restrictions apply to add-on purchases:
+         * - Add-on purchases are currently only supported for subscriptions on the Play Store.
+         * - The renewal periods of all add-on packages must be the same and match the period of the base product.
+         * - No more than 49 add-ons packages per multi-line purchase are allowed.
+         */
+        @ExperimentalPreviewRevenueCatPurchasesAPI
+        @Throws(PurchasesException::class)
```

TBH I’m wondering if having this throw is the best API for this, since Kotlin doesn’t force you to handle it, I can foresee some people would not actually catch this exception…

One alternative I think could be to have a different method in Purchases where it allows you to pass the different packages/products to bundle them and create a new entity for this bundled subscription, this method can return an error. Then, allow to pass this new bundled entity when creating a PurchaseParams...

Wdyt @RevenueCat/coresdk ?

---

Yeah, it's definitely possible (and maybe even likely?) for a developer to not try/catch this function, leading to app crashes.

Perhaps we could do something like:

```kotlin
val subscriptionAddOns = Purchases.sharedInstance.createSubscriptionAddOns(
   baseItem: StoreProduct, 
   storeProducts: List<StoreProduct>
) -> Result<SubscriptionAddOns>

// Developers can handle the result however they'd like, we'll use .getOrNull()
// here for simplicity
subscriptionWithAddOns.getOrNull()?.let {
   purchaseParams.addOns(subscriptionAddOns)            
}

// Make purchase
```

I can think of a few potential issues to this approach:
- It slightly complicates the purchase flow for developers, as they need to make an extra function call and handle the case where the add-ons aren't valid (though that's better than crashing)
- It's possible for a developer to pass in one baseItem to `bundleSubscriptionAddOns`, but then pass a different one in to the purchase params. This could lead to a scenario where the add-ons were validated against an add-on item that isn't actually being purchased.
   - In this case, perhaps we could just return an error in the `purchase()` function?

---

I do think that makes a lot of sense and it's probably cleaner! Let's do that!

---

The changes for this got a little big due to moving around the validations & their tests, plus adding a few new tests, so I opened up a PR with the changes here to make it easier to review: https://github.com/RevenueCat/purchases-android/pull/2706